File: codeforces/448C.cpp
#include <bits/stdc++.h>
using namespace std;

int paint(vector<int>& A, int l, int r) {
    int vert = r - l;
    int hmin = *min_element(A.begin() + l, A.begin() + r);
    for (int i = l; i < r; i++) A[i] -= hmin;
    int x = l, y = l;
    
    if (*max_element(A.begin()+l,A.begin()+r)>0) {
        while (x<r) {
            while (x + 1 < r && A[x] == 0) {x++;y++;}
            while (y + 1 < r && A[y] > 0) y++;
            if (y + 1 == r && A[y] > 0) y++;
            hmin += paint(A, x, y);
            x = ++y;
        }
    }
    return min(vert, hmin);
}

int main() {
    int N;
    cin >> N;
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }

    cout << paint(A,0,N) << endl;

    return 0;
}

File: codeforces/448C.py
N = int(input())
A = [*map(int,input().split())]

def paint(l,r):
	global A
	vert = r-l
	hmin = min(A[l:r])
	for i in range(l,r): A[i]=A[i]-hmin
	x,y = l,l
	if max(A[l:r])>0:
		while x<r:
			while y+1<r and A[y]>0: y+=1
			while x+1<r and A[x]==0: x+=1
			if y+1==r and A[y]>0: y+=1
			hmin += paint(x,y)
			x,y=y,y+1
	return min(vert,hmin)

print(paint(0,N))
